src/auxiliary/p4est.jl

```diff
@@ -0,0 +1,162 @@
+"""
+    init_p4est()
+
+Initialize p4est by calling `p4est_init` and setting the log level to `SC_LP_ERROR`.
+This function will check if p4est is already initialized
+and if yes, do nothing, thus it is safe to call it multiple times.
+"""
+function init_p4est()
+  if p4est_package_id()[] >= 0
+    return nothing
+  end
+
+  # Initialize p4est with log level ERROR to prevent a lot of output in AMR simulations
+  p4est_init(C_NULL, SC_LP_ERROR)
+
+  return nothing
+end
+
+
+# Convert sc_array of type T to Julia array
+function unsafe_wrap_sc(::Type{T}, sc_array) where T
+  element_count = sc_array.elem_count
+  element_size = sc_array.elem_size
+
+  @assert element_size == sizeof(T)
+
+  return [unsafe_wrap(T, sc_array.array + element_size * i) for i in 0:element_count-1]
+end
+
+
+# Load the ith element (1-indexed) of an sc array of type T
+function unsafe_load_sc(::Type{T}, sc_array, i=1) where T
+  element_size = sc_array.elem_size
+
+  @assert element_size == sizeof(T)
+
+  return unsafe_wrap(T, sc_array.array + element_size * (i - 1))
+end
+
+
+function new_p4est(conn::Ptr{p4est_connectivity_t}, initial_refinement_level)
+  p4est_new_ext(0, # No MPI communicator
+                conn,
+                0, # No minimum initial qudrants per processor
+                initial_refinement_level,
+                true, # Refine uniformly
+                2 * sizeof(Int), # Use Int-Vector of size 2 as quadrant user data
+                C_NULL, # No init function
+                C_NULL) # No user pointer
+end
+
+function new_p4est(conn::Ptr{p8est_connectivity_t}, initial_refinement_level)
+  p8est_new_ext(0, conn, 0, initial_refinement_level, true, 2 * sizeof(Int), C_NULL, C_NULL)
+end
+
+
+function save_p4est!(file, p4est::Ptr{p4est_t})
+  # Don't save user data of the quads
+  p4est_save(file, p4est, false)
+end
+
+function save_p4est!(file, p8est::Ptr{p8est_t})
+  # Don't save user data of the quads
+  p8est_save(file, p8est, false)
+end
+
+
+# 2D
+function load_p4est(file, ::Val{2})
+  conn_vec = Vector{Ptr{p4est_connectivity_t}}(undef, 1)
+  p4est_load(file, C_NULL, 0, false, C_NULL, pointer(conn_vec))
+end
+
+# 3D
+function load_p4est(file, ::Val{3})
+  conn_vec = Vector{Ptr{p8est_connectivity_t}}(undef, 1)
+  p8est_load(file, C_NULL, 0, false, C_NULL, pointer(conn_vec))
+end
+
+
+# 2D
+read_inp_p4est(meshfile, ::Val{2}) = p4est_connectivity_read_inp(meshfile)
+# 3D
+read_inp_p4est(meshfile, ::Val{3}) = p8est_connectivity_read_inp(meshfile)
+
+
+refine_p4est!(p4est::Ptr{p4est_t}, recursive, refine_fn_c, init_fn_c) = p4est_refine(p4est, recursive, refine_fn_c, init_fn_c)
+refine_p4est!(p8est::Ptr{p8est_t}, recursive, refine_fn_c, init_fn_c) = p8est_refine(p8est, recursive, refine_fn_c, init_fn_c)
```

Comment:
I started implementing a few proxy functions here that dispatch by dimension to call the correct `p4est` function. For a lack of better naming ideas, I basically just reversed the function name and wrote `p4est` at the end.

Reply:
Are these all mutating functions? If so, you could just use the name with `!` in generic code and call the appropriate functions without `!` from p4est.

Reply:
Note sure I agree with the `p4est_` prefix. Right now, this is strictly reserved to p4est library-defined functions, which is a nice property in terms of code understandability IMHO. What about just using different names, e.g., `read_inp_meshfile` and `refine!`?

Reply:
@ranocha, no, not all.
@sloede, `load_p4est` and `new_p4est` would be a bit weird without the `_p4est`.